spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+The API changes described here refer to the Java API, but this SPICE implies analogous changes to the Swift and Go APIs.
+
+The `EvaluatorBuilder` gets the following new methods:
+
+[source,java]
+----
+public final class EvaluatorBuilder {
+    // ...
+
+    /**
+     * Adds the given URI rewrite rule.
+     *
+     * @param replace   The literal string to match.
+     * @param with      The literal string to substitute for the first match of [replace].
+     */
+    public EvaluatorBuilder addRewriteRule(String replace, String with) {
+        // ...
+    }
+
+    /**
+     * Adds the given URI rewrite rule.
+     *
+     * @param replace   The regular expression to match.
+     * @param with      The string to substitute for the first match of [replace], which may contain
+     *                  references to capture groups (`$0`, `$1`, etc).
+     */
+    public EvaluatorBuilder addRewriteRule(Pattern replace, String with) {
+        // ...
+    }
+}
+----
+
+=== Standard library module `pkl:Project`
+
+Rewrite rules can be defined project-wide by declaring them in the `PklProject` file.
+
+[source,pkl]
+----
+@Since { version = "0.26.0" }
+class UriRewrite {
+  /// What to match in the given URI.
+  replace: String|Regex
+
+  /// What to replace the _all_ matches with.
+  ///
+  /// When `replace` is a `String`, `with` is interpreted literally.
+  /// When `replace` is a `Regex`, `with` is interpreted as a replacement pattern,
+  /// which may contain references to capture groups (`$0`, `$1`, etc).
+  with: String
+}
+
+/// Rewrite rules for URIs used in module keys and resource reads.
+///
+/// For any URI, the first of the longest-matching [UriRewrite]s is applied, before applying the secutiry manager.
+@Since { version = "0.26.0" }
+uriRewrites: Listing<UriRewrite>
+----
+
+=== Common CLI flags
+
+These flags are _common_, because this setting applies to different commands equally.
+The proposed `analyze` command (see link:../spices/0001-import-graph-analyzer.pkl[SPICE 1: Import Graph Analyzer API]), for example, applies these rewrites in the resolution of all URIs analyzed and reported.
+
+This proposal adds two flags to the CLI:
+  * `--uri-rewrite <regex> <replacement>`, which adds a URI rewrite rule; and
+  * `--no-uri-rewrites`, which discards all URI rewrite rules defined thus far.
+
+If a `pkl:Project` file is found, this is read first.
+To invoke Pkl without rewrite rules defined therein, `--no-uri-rewrites` can be used, followed by zero-or-more `--uri-rewrite` arguments.
+
+=== Rewrite behavior
+
+The proposed design is to match any URI being resolved in its entirety to all the provided rewrites' `replace` specification.
+The first, longest-matching rewrite rule is applied.
+In other words:
+ - when multiple rules match, a rule that matches a longer substring of the URI is preferred over one matching a shorter substring; and
+ - when multiple rules match the same substring length, the first rule is chosen (in definition order).
```

Comment:
Since these rewrite rules are user-defined as an in-order list, I think we can just pick the first rewrite rule that matches.

It would also be cheaper this way; it allows us to short-circuit once we find a rule that matches.

Reply:
We might also want to validate that a rewrite can only be rewritten to the same scheme, and perhaps that the rewrite is only for http/https.

Rewriting to a different scheme can be problematic because of minor differences between how different schemes work (e.g. HTTPS urls don't support glob imports).

Reply:
So we want this to intentionally exclude custom resource readers?

Reply:
Would it be potentially possible to supporting rewriting between schemes that share the same hierarchical/globbable(/local?) properties?